compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -613,6 +618,57 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }),
         }
     }
+
+    /// Detect the inference regression on crate `time` <= 0.3.35 and emit a more targeted error.
+    /// <https://github.com/rust-lang/rust/issues/127343>
+    // FIXME: we should figure out a more generic version of doing this, ideally in cargo itself.
+    fn detect_old_time_crate_version(
+        &self,
+        span: Option<Span>,
+        /// We will clear the non-actionable suggestion from the error to reduce noise.
+        infer_subdiags: &mut Vec<SourceKindSubdiag<'_>>,
+    ) -> Option<()> {
+        if self.infcx.tcx.crate_name(LOCAL_CRATE) != sym::time {
+            // Only relevant when building the `time` crate.
+            return None;
+        }
+        let Some(span) = span else { return None };
+        let FileName::Real(file) = self.infcx.tcx.sess.source_map().span_to_filename(span) else {
+            return None;
+        };
+        let path = file.local_path_if_available();
+        let mut components = path.components();
+        // We will take the filename of the error and see if it is of the form
+        // `.../registry/src/index.crates.io-.../time-0...`, in order to detect the specific case
+        // we care about.
+        while let Some(component) = components.next() {
+            let std::path::Component::Normal(component) = component else {
+                continue;
+            };
+            if component == "registry"
+                && let Some(next) = components.next()
+                && let std::path::Component::Normal(next) = next
+                && next == "src"
+                && let Some(next) = components.next()
+                && let std::path::Component::Normal(next) = next
+                && next.to_string_lossy().starts_with("index.")
+                && let Some(next) = components.next()
+                && let std::path::Component::Normal(next) = next
+                && let next = next.to_string_lossy()
+                && next.starts_with("time-0.")
+                && let Some(version) = next.split('-').skip(1).next()
+                && let mut segments = version.split('.')
+                && let Some(major) = segments.next()
+                && major == "0"
+                && let Some(minor) = segments.next()
+                && minor == "3"
```

Comment:
Hard-coding an error message like this based off of path matching can likely be written using regex instead. This let chain is inscrutable to the point that I'd rather we not merge this if it can't be rewritten.

Reply:
Or, alternatively, I'd rather we not do path parsing here and just match on a combination of crate name and some other condition we can deduce from the program itself.

Reply:
like, are we even always going to be compiling `time` from a crates.io registry? Does this work correctly for vendored deps? This certainly doesn't work for non-cargo build tools, right?

Reply:
Pattern matching on a path sounds more brittle. This check is horrible, I agree, but it at least handles paths across windows/unix correctly, and by being this *specific* in the number of checks we're less likely to wrongly trigger if someone just happens to put code in a directory called `time-0.blah`. I think we can likely split some of the elements into one or two closures to make the chain easier to read, though.

Reply:
This *is* already incredibly brittle. See that comment I just left. I'd rather we not make this hack rely on some special structure of the crates.io registry, or not go with this at all if that's impossible.